vm-migration/src/state.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::sync::{Arc, Mutex, RwLock};
+use std::collections::HashMap;
+use std::sync::mpsc::{channel, Sender, SendError, RecvError};
+
+/// MigrationState errors are sent back from the receiver through the ApiResponse.
+#[derive(Debug)]
+pub enum MigrationStateError {
+    /// MigrationRequest send error
+    RequestSend(SendError<MigrationRequest>),
+
+    /// Reponse receive error
+    ResponseRecv(RecvError),
+
+    /// Cannot handle migration state load request.
+    MigrationStateLoad,
+}
+pub type MigrationStateResult<T> = std::result::Result<T, MigrationStateError>;
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct MigrationStateData {
+    pub state: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub enum MigrationResponsePayload {
+    /// No data is sent on the channel.
+    Empty,
+
+    /// Migration state content
+    MigrationState(MigrationStateData),
+}
+
+/// This is the response sent by the VMM API server through the mpsc channel.
+pub type MigrationStateResponse = std::result::Result<MigrationResponsePayload, MigrationStateError>;
+
+#[allow(clippy::large_enum_variant)]
+pub enum MigrationRequest {
+    /// Request to get migratble states. The response payload is a json which
+    /// contains serialized states.
+    /// If the receiver could not correctly handle this request, it will send a
+    /// MigrationStateGet error back.
+    MigrationStateGet(Sender<MigrationStateResponse>),
+
+    /// Request to load migratble states. The request payload is a json which
+    /// contains serialized states.
+    /// If the receiver could not correctly handle this request, it will send a
+    /// MigrationStateLoad error back.
+    MigrationStateLoad(Arc<MigrationStateData>),
+}
+
+lazy_static! {
+    static ref MIGRATION_STATE_ROUTES: Arc<RwLock<HashMap<String, Mutex<Sender<MigrationRequest>>>>> =
+        Arc::new(RwLock::new(HashMap::new()));
+}
+
```

Comment:
I'd like to keep this array in migration component because our final
goal is to release 'vm-migration' as a crate. When registering sender
to this array, the migration instance/thread is not created yet. So,
I declare it as static.

BTW, current mechanism depends Api. Do we have plan to submit it as
a crate? If it only exists in vmm, every new vmm which uses
'vm-migration' crate should implements such mechansim.